core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/kerberos/KeytabConfigurationHandler.java

```diff
@@ -0,0 +1,119 @@
+
+    @Inject
+    private HostOrchestrator hostOrchestrator;
+
+    @Inject
+    private SecretService secretService;
+
+    @Override
+    public String selector() {
+        return EventSelectorUtil.selector(KeytabConfigurationRequest.class);
+    }
+
+    @Override
+    public void accept(Event<KeytabConfigurationRequest> keytabConfigurationRequestEvent) {
+        Long stackId = keytabConfigurationRequestEvent.getData().getResourceId();
+        Selectable response;
+        try {
+            Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+            Optional<KerberosConfig> kerberosConfigOptional = kerberosConfigService.get(stack.getEnvironmentCrn());
+            // TODO remove YARN check when FreeIPA registration is ready
+            if (!CloudPlatform.YARN.name().equals(stack.cloudPlatform())
+                    && kerberosConfigOptional.isPresent() && kerberosDetailService.isIpaJoinable(kerberosConfigOptional.get())) {
+                GatewayConfig primaryGatewayConfig = gatewayConfigService.getPrimaryGatewayConfig(stack);
+                ServiceKeytabResponse serviceKeytabResponse = getServiceKeytabResponse(stack, primaryGatewayConfig);
+                KeytabModel keytabModel = buildKeytabModel(serviceKeytabResponse);
+                hostOrchestrator.uploadKeytabs(List.of(primaryGatewayConfig), Set.of(keytabModel),
+                        ClusterDeletionBasedExitCriteriaModel.clusterDeletionBasedModel(stackId, stack.getCluster().getId()));
+            }
+            response = new KeytabConfigurationSuccess(stackId);
+        } catch (Exception e) {
+            LOGGER.info("Error during keytab configuration, stackId: " + stackId, e);
+            response = new KeytabConfigurationFailed(stackId, e);
+        }
+        eventBus.notify(response.selector(), new Event<>(keytabConfigurationRequestEvent.getHeaders(), response));
+    }
+
+    private KeytabModel buildKeytabModel(ServiceKeytabResponse serviceKeytabResponse) {
+        String keytabInBase64 = secretService.getByResponse(serviceKeytabResponse.getKeytab());
+        byte[] keytab = Base64.getDecoder().decode(keytabInBase64.getBytes(StandardCharsets.UTF_8));
+        String principal = secretService.getByResponse(serviceKeytabResponse.getServicePrincial());
+        return new KeytabModel("CM", "/etc/cloudera-scm-server", "cmf.keytab", principal, keytab);
+    }
+
+    private ServiceKeytabResponse getServiceKeytabResponse(Stack stack, GatewayConfig primaryGatewayConfig)
+            throws Exception {
+        ServiceKeytabRequest request = new ServiceKeytabRequest();
+        request.setEnvironmentCrn(stack.getEnvironmentCrn());
+        request.setServerHostName(primaryGatewayConfig.getHostname());
+        request.setServiceName("CM");
+        request.setDoNotRecreateKeytab(Boolean.TRUE);
+        RoleRequest roleRequest = new RoleRequest();
+        roleRequest.setRoleName("hadoopadminrole-" + stack.getName());
```

Comment:
Same comment as above.

Reply:
it's still internal stuff for CB. We use these to replace this script for generating keytabs: https://github.com/hortonworks/cloudbreak/blob/master/orchestrator-salt/src/main/resources/salt/salt/sssd/template/generate_cm_keytab.j2

Reply:
I understand that this code was added to address the generation of CM keytab but there will be other cases where we would want to use KeytabConfigurationHandler to create keytab for other services.

Reply:
I don't get it. How would you use this handler. This is cloudbreak code.

Reply:
I'm not aware of any other usecase where CB needs a keytab. If it happens we will have to rewrite this piece of code